src/pru-motion-queue.cc

```diff
@@ -82,54 +82,18 @@ static void DumpMotionSegment(volatile const struct MotionSegment *e,
 }
 #endif
 
-// Store the required informations needed to backtrack the absolute position.
-struct HistorySegment {
-  HistorySegment() : fractions(), cumulative_loops(), direction_bits(0) {}
-  uint32_t fractions[MOTION_MOTOR_COUNT];
-  uint32_t cumulative_loops[MOTION_MOTOR_COUNT];
-  uint8_t direction_bits;
-};
-
-void PRUMotionQueue::RegisterHistorySegment(const MotionSegment &element) {
-  const uint64_t max_fraction = 0xFFFFFFFF;
-  const unsigned int last_insert_index = (queue_pos_ - 1) % QUEUE_LEN;
-  const struct HistorySegment &previous
-    = shadow_queue_[(last_insert_index - 1) % QUEUE_LEN];
-
-  const uint64_t total_loops
-    = element.loops_accel + element.loops_travel + element.loops_decel;
-  const uint8_t direction_bits = element.direction_bits;
-
-  HistorySegment *new_slot = &shadow_queue_[last_insert_index];
-
-  // TODO: Motor operations already holds this information, we should wire it
-  // until here in order to avoid this redundant operation.
-  for (int i = 0; i < MOTION_MOTOR_COUNT; ++i) {
-    const uint64_t fraction = element.fractions[i];
-    new_slot->fractions[i] = fraction;
-    const int sign = (direction_bits >> i) & 1 ? -1 : 1;
-    uint64_t loops = fraction * total_loops;
-    loops /= max_fraction;
-    new_slot->cumulative_loops[i] = previous.cumulative_loops[i]
-                                    + sign * loops;
-  }
-
-  new_slot->direction_bits = element.direction_bits;
-}
-
-void PRUMotionQueue::GetMotorsLoops(MotorsRegister *absolute_pos_loops) {
-  const uint64_t max_fraction = 0xFFFFFFFF;
+void PRUMotionQueue::Status(uint32_t *executed_loops,
+                            unsigned int *buffer_size) {
+  // Get data from the PRU
   const struct QueueStatus status = *(struct QueueStatus*) &pru_data_->status;
-  const struct HistorySegment &current = shadow_queue_[status.index];
-
-  const uint64_t counter = status.counter;
-  for (int i = 0; i < MOTION_MOTOR_COUNT; ++i) {
-    const int64_t sign = (current.direction_bits >> i) & 1 ? -1 : 1;
-    uint64_t loops = current.fractions[i];
-    loops *= counter;
-    loops /= max_fraction;
-    (*absolute_pos_loops)[i] = current.cumulative_loops[i] - sign * loops;
+  if (executed_loops) *executed_loops = status.counter;
+  // Nothing has been pushed yet.
+  if (status.index == 0xff) {
+    *buffer_size = 0;
+    return;
   }
+  *buffer_size = (queue_pos_ - status.index) % QUEUE_LEN;
+  *buffer_size += !(*buffer_size) * QUEUE_LEN;
```

Comment:
Yes, sometimes such tricks can result in not having a branch. But
  * Never do premature optimization. Only ever consider manual optimization in code that has shown to be too slow in a critical path.
  * If you do things like this: _measure its impact_. Some (most?) well meaning manual optimizations can be worse (e.g. here you replace a MOV with a multiplication which typically is more expensive, so a positive impact has to be proven - and also is platform dependent).
  * If you do "clever" things that are not necessarily obvious for the reader, you have to comment and explain them, including reasons for need of optimization and measurement results of the impact.
  * In general: let the compiler be "clever". If it sees a ?: operation, it can do much better optimization (e.g. if it knows that multiplication is cheap, it can do what you did automatically). If you obfuscate your intend, it is harder for the compiler to do its job.